test/test_torch.py

```diff
@@ -5483,6 +5483,98 @@ def test_dot_data_use(self):
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_3d, out=output), result), True)
+            result = torch.tensor([[[1, 3, 4], [0, 2, 4]], [[1, 1, 2], [2, 2, 3]]], device=device)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_3d, right=True), result), True)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_3d, right=True, out=output), result), True)
+
+            # simple 1d boundary and 3d input value
+            boundaries = torch.tensor([1, 2, 3, 4, 5, 6], device=device)
+            result = torch.tensor([[[0, 2, 4], [1, 3, 5]], [[0, 1, 2], [3, 4, 5]]], device=device)
+            output = torch.zeros(2, 2, 3, device=device, dtype=torch.int64)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_3d), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_3d, boundaries), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_3d, boundaries, out=output), result), True)
+            result = torch.tensor([[[1, 3, 5], [2, 4, 6]], [[1, 2, 3], [4, 5, 6]]], device=device)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_3d, right=True), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_3d, boundaries, right=True), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_3d, boundaries, out=output, right=True), result), True)
+
+            # simple float 1d boundary and 1d input with output int32 type
+            values_1d_float = values_1d.to(torch.float32)
+            boundaries = torch.tensor([0.9, 1, 2, 2, 3, 3, 4, 4.1, 9, 9], device=device, dtype=torch.float32)
+            result = torch.tensor([1, 2, 4, 6, 8, 8, 8, 8, 8], device=device, dtype=torch.int32)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_1d_float, out_int32=True), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_1d_float, boundaries, out_int32=True), result), True)
+
+            # multiple dimension input with 0 elements
+            boundaries = torch.tensor([1, 2, 3, 4, 5, 6], device=device, dtype=torch.int64)
+            values_0_el = torch.tensor([[[]]], device=device, dtype=torch.int64)
+            result = values_0_el.to(torch.int64)
+            self.assertEqual(torch.equal(torch.searchsorted(boundaries, values_0_el), result), True)
+            self.assertEqual(torch.equal(torch.bucketize(values_0_el, boundaries), result), True)
+
+            # nan input
+            values_nan = torch.tensor([1.0, float('nan'), 2.0, float('nan')], device=device, dtype=torch.float64)
+            boundaries = torch.tensor([0.0, 1.0, 2.0, 3.0], device=device, dtype=torch.float64)
+            result_left = torch.tensor([1, 0, 2, 0], device=device)
+            result_right = torch.tensor([2, 4, 3, 4], device=device)
```

Comment:
This won't do, nans always should output the last element, regardless of whether it's right or left.

Reply:
I see. I've thought this is the same as std::upper_bound and lower_bound. Will correct the code.